### DISCUSSION TITLE
Why there is a difference in accuracy score

### DISCUSSION THREAD

**Discussion by ogarnijAI** [rasbt/mlxtend#860] [Q&A]
Hello,
I am using an example code related to ExhaustiveFeatureSelector. 

```
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.33)

from sklearn import tree
%matplotlib qt

clf = tree.DecisionTreeClassifier(criterion="entropy",random_state=42) #,max_depth=3

efs1 = EFS(clf, 
           min_features=4,
           max_features=8,
           scoring='accuracy',
           print_progress=True,
           cv=5,
           n_jobs=-1,
           pre_dispatch=15 
           )

efs1 = efs1.fit(X_train, y_train)

print('Best accuracy score: %.2f' % efs1.best_score_)
print('Best subset (indices):', efs1.best_idx_)
print('Best subset (corresponding names):', efs1.best_feature_names_)
```
The result is:
Features: 64702/64702
Best accuracy score: **0.93**
Best subset (indices): (0, 6, 12, 16)

then I do that:
```
X_train_efs = efs1.transform(X_train)
X_test_efs = efs1.transform(X_test)
clf.fit(X_train_efs, y_train)
y_pred = clf.predict(X_test_efs)

# Compute the accuracy of the prediction
acc = float((y_test == y_pred).sum()) / y_pred.shape[0]
print('Test set accuracy: %.2f %%' % (acc*100))
```
and got: Test set accuracy: 64.71 %

Why such a difference in result ? I would accept if this 0.93 would be the best score in CV, it can happen for an overfitted tree. But the documentation says:

_best_score_ : float
Cross validation average score of the selected subset._

Any ideas ?

**Comment by rasbt** [comment]
That's a good point. Maybe there's something weird about the test set. Can you try

```python
sklearn.model_selection.cross_val_score


scores = cross_val_score(clf, X_train_efs, y_train, cv=5)
print(np.mean(scores)*100)
```

to see if you get something in the 93% range?

**Reply by ogarnijAI** [reply]
```
from sklearn.model_selection import cross_val_score

scores = cross_val_score(clf, X_train_efs, y_train, cv=5)
print(scores)
print(np.mean(scores)*100)

```

[1.         1.         0.83333333 1.         0.83333333]
93.33333333333333

this is weird ... the entire dataset is only 263 samples, but I did quite a few tests and for a limited set of features I got eg.:

------------------------------------------------------------
              Train data:
------------------------------------------------------------
[[97 15]
 [20 78]]
              precision    recall  f1-score   support

           0       0.83      0.87      0.85       112
           1       0.84      0.80      0.82        98

    accuracy                           0.83       210
   macro avg       0.83      0.83      0.83       210
weighted avg       0.83      0.83      0.83       210

Accuracy : 83.333%
------------------------------------------------------------
              Test data :
------------------------------------------------------------
[[22  9]
 [12 10]]
              precision    recall  f1-score   support

           0       0.65      0.71      0.68        31
           1       0.53      0.45      0.49        22

    accuracy                           0.60        53
   macro avg       0.59      0.58      0.58        53
weighted avg       0.60      0.60      0.60        53

Accuracy : 60.377%
------------------------------------------------------------
              Full data :
------------------------------------------------------------
[[119  24]
 [ 32  88]]
              precision    recall  f1-score   support

           0       0.79      0.83      0.81       143
           1       0.79      0.73      0.76       120

    accuracy                           0.79       263
   macro avg       0.79      0.78      0.78       263
weighted avg       0.79      0.79      0.79       263

Accuracy : 78.707%
------------------------------------------------------------
     Cross validation:
------------------------------------------------------------
Accuracy: 0.67 (+/- 0.08)
Scores  : [0.73584906 0.69811321 0.64150943 0.63461538 0.65384615]

but never 1 or 0.93.

**Comment by rasbt** [comment]
Hm, that's strange. The fact that you get similar results with cross_val_score hints that at least it's probably not due to overfitting. Given that this is such a small dataset, I am wondering whether a larger k would be better. Or maybe leave one out cross validation

**Comment by ogarnijAI** [comment]
k=7
```
[1.   1.   0.8  1.   1.   1.   0.75]
93.57142857142857
```

k=10
```
[1.         1.         1.         1.         1.         1.
 1.         0.66666667 1.         0.66666667]
93.33333333333333
```
k=15
```
[1.  1.  1.  1.  1.  1.  1.  1.  1.  1.  1.  1.  0.5 1.  0.5]
93.33333333333333
```

**Reply by rasbt** [reply]
The fact that things fluctuate so much within each fold might be a hint that there is something weird with the dataset. If you change the random seed for the training/test split, this would probably could also widely change the test accuracy.

**Comment by TheSleepingDragon** [comment]
Have you tried the stratify feature in `train_test_split` to test whether stratification could improve the results or not (or just several other random_states for train_test_split? Because it seems that your K folds in cross-validation are fine.
